### DISCUSSION TITLE
MIRYOKU ZMK: How to create Mod Morph keys for "DOT PLUS" and "COMMA MINUS" custom keys

### DISCUSSION THREAD

**Discussion by fsqonbdk** [manna-harbour/miryoku#283] [Q&A] [answered]
Hi,

I would like to make it so that they key that currently is assigned DOT (.) has the behavior of DOT when unmodified and PLUS (+) when shifted. In ZMK I believe the way to do this is Mod Morph, something like this:

/ { \
  behaviors { \
    dotplus: dot_plus { \
      compatible = "zmk,behavior-mod-morph"; \
      label = "DOTPLUS"; \
      #binding-cells = <0>; \
      bindings = <&kp DOT>, <&kp PLUS>; \
      mods = <(MOD_LSFT|MOD_RSFT)>; \
    }; \
  }; \
};

I then replaced my "DOT" in my custom_config.h layout, from U_MT(RALT, DOT) to U_MT(RALT, DOTPLUS) and got an error.

My custom_config.h snippet showing layout:

#define MIRYOKU_LAYER_BASE \
&kp Z,       &kp L,             &kp H,             &kp W,             &kp K,             &kp J,             &kp F,             &kp O,              &kp U,             &kp COMMA,          \
U_MT(LGUI, S),     U_MT(LALT, R),     U_MT(LCTRL, N),    U_MT(LSHFT, T),    &kp B,             &kp Y,             U_MT(LSHFT, C),    U_MT(LCTRL, E),     U_MT(LALT, I),     U_MT(LGUI, A),      \
U_LT(U_BUTTON, V), U_MT(RALT, X),     &kp M,             &kp D,             &kp Q,             &kp G,             &kp P,             &kp SQT,            U_MT(RALT, DOTPLUS),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, ESC),  U_LT(U_NUM, SPACE),U_LT(U_SYM, TAB),  U_LT(U_MOUSE, RET),U_LT(U_NAV, DEL),  U_LT(U_MEDIA, BSPC),U_NP,              U_NP

I realize I may be asking two questions:

1. How to implement a simple Mod Morph key in Miryoku.
2. How to implement a Mod Morph key inside U_MT, which may require creating a custom behavior such as U_MT_DOTPLUS.

Thank you for your amazing work and for your help.

**Comment by fsqonbdk** [comment, marked as answer]
I believe I have it figured out.

Here is how I solved question 1 of 2:

Steps taken:
1. create custom_keymap.dtsi in miryoku directory (same location as custom_config.h). Content of custom_keymap.dtsi is 

/ { \
  behaviors { \
    dotplus: dotplus { \
      compatible = "zmk,behavior-mod-morph"; \
      label = "DOTPLUS"; \
      #binding-cells = <0>; \
      bindings = <&kp DOT>, <&kp PLUS>; \
      mods = <(MOD_LSFT|MOD_RSFT)>; \
    }; \
  }; \
};

2. Add the following line to miryoku.dtsi:

#include "custom_keymap.dtsi"

3. Update custom_config.h to use newly created Mod Morph key &dotplus (I put it in place of the 1st key, upper left for testing, not its final location):

#define MIRYOKU_LAYER_BASE \
&dotplus,             &kp L,             &kp H,             &kp W,             &kp K,             &kp J,             &kp F,             &kp O,              &kp U,             &kp COMMA,          \
U_MT(LGUI, S),     U_MT(LALT, R),     U_MT(LCTRL, N),    U_MT(LSHFT, T),    &kp B,             &kp Y,             U_MT(LSHFT, C),    U_MT(LCTRL, E),     U_MT(LALT, I),     U_MT(LGUI, A),      \
U_LT(U_BUTTON, V), U_MT(RALT, X),     &kp M,             &kp D,             &kp Q,             &kp G,             &kp P,             &kp SQT,            U_MT(RALT, DOT),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, ESC),  U_LT(U_NUM, SPACE),U_LT(U_SYM, TAB),  U_LT(U_MOUSE, RET),U_LT(U_NAV, DEL),  U_LT(U_MEDIA, BSPC),U_NP,  

This compiles and works on physical keyboard (corne 36 nice nano v2).

For question 2 of 2, I appear to have this working by adding a new behavior to use this Mod Morph key in a U_MT or U_LT key, similar to what is described in this post: [Combing ModMorph and Layer Tap](https://www.reddit.com/r/zmk/comments/12ujlmb/combing_modmorph_and_layer_tap/).

Here are the steps I took in addition to what I did above:

1. added new behavior, &u_mtdp (mod tap dot plus) to custom_keymap.dtsi:

/ {
  behaviors {
    dotplus: dotplus {
      compatible = "zmk,behavior-mod-morph";
      label = "DOTPLUS";
      #binding-cells = <0>;
      bindings = <&kp DOT>, <&kp PLUS>;
      mods = <(MOD_LSFT|MOD_RSFT)>;
    };
    u_mtdp: u_mtdp {
      compatible = "zmk,behavior-hold-tap";
      label = "u_mtdp";
      #binding-cells = <2>;
      tapping_term_ms = <U_TAPPING_TERM>;
      flavor = "tap-preferred";
      bindings = <&kp>, <&dotplus>;
    };
  };
};

2. Created new file custom_keymap.h

// Copyright 2022 Manna Harbour
// https://github.com/manna-harbour/miryoku

#pragma once

#define U_MTDP(MOD, TAP) &u_mtdp MOD 0

3. Added line below to miryoku.dtsi

#include "custom_keymap.dtsi"

4. Added line below to miryoku.h

#include "custom_keymap.h"

5. Updated my custom keymap in custom_config.h to use U_MTDP(RALT, DOTPLUS)

#define MIRYOKU_LAYER_BASE \
&kp Z,             &kp L,             &kp H,             &kp W,             &kp K,             &kp J,             &kp F,             &kp O,              &kp U,             &kp COMMA,          \
U_MT(LGUI, S),     U_MT(LALT, R),     U_MT(LCTRL, N),    U_MT(LSHFT, T),    &kp B,             &kp Y,             U_MT(LSHFT, C),    U_MT(LCTRL, E),     U_MT(LALT, I),     U_MT(LGUI, A),      \
U_LT(U_BUTTON, V), U_MT(RALT, X),     &kp M,             &kp D,             &kp Q,             &kp G,             &kp P,             &kp SQT,            U_MTDP(RALT, DOTPLUS),   U_LT(U_BUTTON, SLASH),\
U_NP,              U_NP,              U_LT(U_FUN, ESC),  U_LT(U_NUM, SPACE),U_LT(U_SYM, TAB),  U_LT(U_MOUSE, RET),U_LT(U_NAV, DEL),  U_LT(U_MEDIA, BSPC),U_NP,              U_NP

And it now compiles and works as desired!

Did I do this correctly? Is there a more elegant or general way I should be implementing this to be consistent with how you have structured Miryoku?

Thanks again for your help.

**Reply by manna-harbour** [reply]
Glad you worked it out, and that all looks good!

Using other behaviors with tap hold requires these kind of DT gymnatics, and this a good implementation.

You could just append to `miryoku_behaviors.dtsi` and `miryoku_behaviors.h`, but adding `custom_keymap.dtsi` and `custom_keymap.h` and the corresponding includes is cleaner, and those files will be added to Miryoku ZMK in the future for this purpose.

Adding the `U_MTDP` macro isn't really necessary and you could just add the `&u_mtdp` behavior to the layer directly, but it isn't harmful. (The original `U_MT` macro is mostly just an artifact of the code generation via Miryoku Babel.)

For a single key this is best, but if you want to do it for many you could use some macros as is done in https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_shift_functions.dtsi or https://github.com/manna-harbour/miryoku_zmk/blob/master/miryoku/miryoku_kludge_tapdelay.h.
